app/src/main/java/com/example/smsrc/sms/handlers/SMSExecutor.java

```diff
@@ -1,10 +1,60 @@
 package com.example.smsrc.sms.handlers;
 
+import android.content.Context;
+
+import com.example.smsrc.commands.model.Command;
+import com.example.smsrc.commands.model.CommandFactory;
+import com.example.smsrc.commands.model.CommandsContract;
+import com.example.smsrc.permissions.models.Authenticate;
+import com.example.smsrc.permissions.models.Authorize;
+import com.example.smsrc.permissions.utils.Crypto;
 import com.example.smsrc.sms.model.SMS;
+import com.example.smsrc.users.dals.UserRepository;
+import com.example.smsrc.users.models.User;
+
+import java.util.List;
 
 public class SMSExecutor {
 
+    private UserRepository repository;
+    private Authorize authorize;
+    public void setRepository(UserRepository userRepository) {
+        repository = userRepository;
+        authorize = new Authorize();
+    }
+
+
     public void execute(SMS sms) {
-        // TODO
+
+        String hash;
+        User user = null;
+        List<User> allUsers = repository.getAllUsers();
+        for (User u: allUsers) {
+
+            hash = Crypto.encrypt(u.getUsername() + sms.getRandomness() + u.getPasscode());
+            if(hash.equals(sms.getCredentials())) {
+                user = u;
+                break;
+            }
+        }
+
+        if(user == null)
+            throw new RuntimeException("Credentials not found");
+
+        Command command = null;
+        for (String commandName: CommandsContract.allCommands) {
+            hash = Crypto.encrypt(commandName + sms.getRandomness());
+            if(hash.equals(sms.getCommand())) {
+                if (authorize.authorize(user, commandName)) {
+                    command = CommandFactory.getCommand(commandName);
+                } else
+                    throw new RuntimeException("UnAuthorized command");
+            }
+        }
```

Comment:
Don't you think this code block can be reused when sending a new SMS? If so can you extract it in Crypto

Reply:
it wiil be only 1 line
            line = Crypto.encrypt(sms.getCommand() + sms.getRandomness());
no need to add a new function